pkg/proctree/process.go

```diff
@@ -0,0 +1,375 @@
+package proctree
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/aquasecurity/tracee/pkg/utils/types"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+// processNode is a node in the process tree representing a process.
+// Its purpose is to keep the information of the process, and the connections to other processes
+// and threads nodes.
+// The current implementation is using mutex, so all internal members are assumed to be protected
+// by it (so are not necessarily thread-safe types)
+type processNode struct {
+	// TODO: Add information about the processes like opened files,
+	//  network activities (like TCP connections), argv, environment variables,
+	//  loader and interpreter
+	id             int
+	creationTime   int
+	nsId           int
+	userId         int
+	containerId    string
+	genInfoLock    sync.Once
+	execInfo       *types.TimeSeries[procExecInfo]
+	forkTime       types.Timestamp
+	exitTime       types.Timestamp
+	parentProcess  *processNode
+	childProcesses map[int]*processNode
+	threads        map[int]*threadNode
+	mutex          sync.RWMutex // Protection on all accesses to the process, except for PID reading
+}
+
+// procExecInfo is the information about a process which is changed upon execution
+type procExecInfo struct {
+	Cmd             []string
+	ExecutionBinary FileInfo
+}
+
+func newProcessNode(id int) (*processNode, error) {
+	return &processNode{
+		id:             id,
+		creationTime:   int(time.Now().UnixNano()),
+		threads:        make(map[int]*threadNode),
+		childProcesses: make(map[int]*processNode),
+		execInfo:       types.NewTimeSeries[procExecInfo](procExecInfo{}),
+	}, nil
+}
+
+// setGenInfoFromEventProtected fill the general info of the process (information
+// of the event given by every process received from it) in a way that it is filled
+// only once in an efficient way to reduce performance penalty.
+// This method uses the process lock to reduce unnecessary locking if not needed. Make sure to unlock
+// the lock before using this.
+func (p *processNode) setGenInfoFromEventProtected(event *trace.Event) {
```

Comment:
nit: usually Gen is used in place of generate, which may be confusing in this case.
Consider using general instead